### PR TITLE
[modern browser] logout current user if broswer is not modern

### PR DESCRIPTION
**Pull request by macejmic**
**What this PR does / why we need it**:

> Replace this with a quick explanation of what's included here. 

**Which issue(s) this PR fixes** 

https://issues.jboss.org/browse/THREESCALE-2632

**Verification steps** 

> Replace this with a list of necessary steps to verify the PR meets your expectations.
> e.g. `1. Go to Admin Portal, make sure that ...`

**Special notes for your reviewer**:

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -88,6 +89,21 @@ def sublayout
...
+    if current_account && current_account.provider_can_use?(:modern_browser_check) && browser_not_modern?
+      logout_keeping_session!
+      flash.now[:error] = 'To protect you, we cannot let you use the browser you are using. Please upgrade your browser and login in again.'
+      redirect_to provider_admin_path
+    end
+  end
+
+  def browser_not_modern?
+    browser = Browser.new(request.env['HTTP_USER_AGENT'])
+    browser.known? && !browser.modern?
```
Comment (macejmic) [inline app/controllers/application_controller.rb]: not sure about this, shall we logout user if the browser is unknown?

Comment (hallelujah) [reply app/controllers/application_controller.rb]: Probably not :)

Comment (macejmic) [reply app/controllers/application_controller.rb]: agree.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -88,6 +89,22 @@ def sublayout
 
   protected
 
+  def check_browser
+    request_format = request.format
+    return true if request_format.xml? || request_format.json?
+
+    if current_account&.provider_can_use?(:modern_browser_check) && browser_not_modern?
+      logout_keeping_session!
+      flash.now[:error] = 'To protect you, we cannot let you use the browser you are using. Please upgrade your browser and login again.'
```
Comment (thomasmaas) [inline app/controllers/application_controller.rb]:
```suggestion
      flash.now[:error] = "The browser you are using doesn't seem to support the X-Frame-Options header. That means we can't protect you against Cross Frame Scripting and thus not guarantee the security of your session. Please upgrade your browser and login again."
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -88,6 +89,22 @@ def sublayout
 
   protected
 
+  def check_browser
+    request_format = request.format
+    return true if request_format.xml? || request_format.json?
+
+    if current_account&.provider_can_use?(:modern_browser_check) && browser_not_modern?
```
Comment (thomasmaas) [inline app/controllers/application_controller.rb]: I think we should keep the number of switches to a minimum as it only leads to edge cases and complexity. Maybe just enable for all? What do you think @hallelujah @guicassolato?

Comment (hallelujah) [reply app/controllers/application_controller.rb]: Agree, we can enable for all yes.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (kevprice83) [reply app/controllers/application_controller.rb]: Where are we configuring the rules for the modern browser compatibility? Right now it looks like we are just using the defaults from the library which are not sane at all. Look at the firefox, opera & IE defaults, they are years old. If we are setting this somewhere else where those versions are relevant then ignore this comment and please point me to that commit. We should list this in our supported configurations.

Comment (kevprice83) [reply app/controllers/application_controller.rb]: Maybe we can use [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options#Browser_compatibility) as a reference?

Comment (macejmic) [reply app/controllers/application_controller.rb]: @kevprice83 defaults from the library mirror X-frame-Options support, taken from: https://caniuse.com/#feat=x-frame-options

Comment (kevprice83) [reply app/controllers/application_controller.rb]: Okay makes sense, the only one that maybe needs review is firefox as the default in the browser gem is [17](https://github.com/fnando/browser/blob/v2.6.1/lib/browser/browser.rb#L87) but in https://caniuse.com/#feat=x-frame-options it is 18. Once we have the final list then I'll speak with Andreu to ensure our supported browser list is up-to-date. Thanks @macejmic

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -88,6 +89,22 @@ def sublayout
...
+    return true if request_format.xml? || request_format.json?
+
+    if current_user && browser_not_modern?
+      logout_keeping_session!
+      flash.now[:error] = "The browser you are using doesn't seem to support the X-Frame-Options header. That means we can't protect you against Cross Frame Scripting and thus not guarantee the security of your session. Please upgrade your browser and sign in again."
+      redirect_to provider_admin_path
+    end
+  end
+
+  def browser_not_modern?
```
Comment (hallelujah) [inline app/controllers/application_controller.rb]: I'd use the positive  `modern_browser?` here

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -88,6 +89,22 @@ def sublayout
 
   protected
 
+  def check_browser
+    request_format = request.format
+    return true if request_format.xml? || request_format.json?
```
Comment (hallelujah) [inline app/controllers/application_controller.rb]: No need of returning true right?
```ruby
return if modern_browser? || request_format.xml? || request_format.json?
```

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (macejmic) [reply app/controllers/application_controller.rb]: we would need to return true if a browser is unknown.

Comment (macejmic) [reply app/controllers/application_controller.rb]: taking into account that, I would prefer `modern_not_modern?`

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (macejmic) [reply app/controllers/application_controller.rb]: true.

**Comment by macejmic** [comment]
@kevprice83 https://github.com/3scale/porta/pull/1123/files#diff-40a9819bb61b5a4203d5bb7c2c5fdde2

**Pull request merged**